src/main/java/uk/gov/hmcts/reform/sendletter/tasks/UploadLettersTask.java

```diff
@@ -63,27 +64,39 @@ public void run() {
 
         // Upload the letters in batches.
         // With each batch we mark them Uploaded so they no longer appear in the query.
-        List<Letter> letters;
-        int counter = 0;
-
-        do {
-            letters = repo.findFirst10ByStatus(LetterStatus.Created);
-            letters
-                .forEach(letter -> {
-                    uploadToFtp(letter);
-                    markAsUploaded(letter);
-                });
-            counter += letters.size();
-        } while (!letters.isEmpty());
-
-        if (counter > 0) {
-            insights.trackUploadedLetters(counter);
+        List<Letter> lettersToUpload = repo.findFirst10ByStatus(LetterStatus.Created);
+
+        if (!lettersToUpload.isEmpty()) {
```

Comment:
The way it works is pretty optimal - we read letters once and we open the connection only when there's something to upload.

How about:
```
while (!lettersToUpload.isEmpty()) {
   ...
   lettersToUpload = repo.findFirst10ByStatus(LetterStatus.Created);
}
```

This way we'd get rid of `letters` and some other bits of code.

Reply:
`lettersToUpload` will become dynamic value and lambda function will not appreciate that. Would be nice to have some streaming workaround but I think it is only available from Java 9. On the other hand it will still require to fetch first batch in order to check whether worth opening ftp connection or not.

Reply:
Right, I forgot that it's wrapped in a function.